File: solutions/5-longest-palindromic-substring/longest-palindromic-substring.cpp
class Solution {
public:

    int dp[1001][1001];
    string s;
    int rec(int i, int j){
        if(i>j){
            return 0;
        }
        if(i==j){
            return 1;
        }
        if(dp[i][j]!=-1){
            return dp[i][j];
        }
        int ans=INT_MIN;
        if(s[i]==s[j]){
            ans=max(ans,2+rec(i+1,j-1));
        }
        return dp[i][j]=ans;
    }

    string longestPalindrome(string str) {
        s=str;
        int n=s.length();
        memset(dp,-1,sizeof(dp));
        for(int i=0;i<n;i++){
            dp[i][i]=1;
        }
        for(int i=0;i<n-1;i++){
            for(int j=i+1;j<n;j++){
                if(s[i]==s[j]){
                    rec(i,j);
                }
            }
        }
        int ans=INT_MIN,start,end;
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(ans<dp[i][j]){
                    ans=dp[i][j];
                    start=i;
                    end=j;
                }
            }
        }
        return s.substr(start,end-start+1);
    }
};

File: solutions/Strings/0005-longest-palindromic-substring/longest-palindromic-substring.java
class Solution {
    Boolean dp[][];
    Boolean rec(int i, int j, String s){
        if(i>=j){
            return true;
        }
        if(dp[i][j]!=null){
            return dp[i][j];
        }
        if(s.charAt(i)==s.charAt(j)){
            return dp[i][j]=rec(i+1,j-1,s);
        }
        return dp[i][j]=false;
    }

    public String longestPalindrome(String s) {
        int n=s.length();
        int[] ends={0,0};
        int ans=Integer.MIN_VALUE;
        dp=new Boolean[n+1][n+1];
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(rec(i,j,s) && ans<(j-i+1)){
                    ends[0]=i;
                    ends[1]=j;
                    ans=j-i+1;
                }
            }
        }
        return s.substring(ends[0],ends[1]+1);
    }
}